test/assignment-2/BasicVault.t.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "forge-std/test.sol";
+import "src/assignment-2/BasicVault.sol";
+import "./ERC20MockWithFailedTransfers.sol";
+
+abstract contract ZeroState is Test {
+    BasicVault public v;
+    ERC20MockWithFailedTransfers token;
+
+    function setUp() public virtual {
+        token = new ERC20MockWithFailedTransfers("Walkie Talkie", "WOKTOK");
+        v = new BasicVault(token);
+    }
+}
+
+contract ZeroStateTest is ZeroState {
+
```

Comment:
These tests are looking good in general.  It's clear you've got a good understanding of foundry testing and also the whole "state inheritance" pattern we talked about.

I have to shut down for the day but I will leave a general comment I have is it seems like you might have too many layers, making it overly complex and a bit hard to follow.  Not sure if you need the UserApproved and UserWithTokens layers and associated tests.  Also, the some of the tests you wrote are for the ERC20 code which, for this assignment, you can assume are fully tested.  I want to look at these more closely and give you better feedback, but that is my initial impression.  I like the `failTransfers`, thats nice!